### PR TITLE
introduce clang format

### LINKED ISSUE #562: discuss: introduce clang format
The code is sometimes inconsistently formatted. That does not bother me.

However, when developing, it is very convenient if one can use auto formatting. I also think it leads to cleaner pull requests since unrelated white space changes are less likely to occur. It is also easier for contributors to adhere to the code style without effort (most IDE-s support formatting).

This comes up now when [working with the span interface ](https://github.com/simdutf/simdutf/issues/557)- it would be nice to have some kind of machine enforceable formatting in place before I start.

The established practice is to have a .clang-format checked in at the root level. There is already one for the [fuzzer subdirectory](https://github.com/simdutf/simdutf/blob/master/fuzz/.clang-format).

The downsides: 
 - the project has to agree on a style. I mostly don't mind which one.
 - (proposed, but debatable): making one big reformat commit, which might inconvenience people doing git blame. [there seems to be a handy way to ignore it](https://akrabat.com/ignoring-revisions-with-git-blame/)
 - care has to be taken for [header reordering](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#sortincludes) - this should be disabled in the .clang-format file. otherwise amalgamation and more may break (this project is somewhat special in this regard). This is turned off in the fuzzer subdirectory.
 - sometimes clang format can give inconsistencies. a recent example is handling concepts, which changed in recent versions. considering the codebase is mostly C++11, I don't think this is a problem.

I suggest to introduce a CI job which checks the formatting. In case there is a formatting failure, the job fails and a suggested diff is output which can be applied with git apply.

The suggestion is:
- add a .clang-format file, (perhaps the one in the fuzzer subdir)
- add a CI job
- add a reformatting commit
- add a .git-blame-ignore-revs  file containing the reformatting commit

### PR DESCRIPTION
**Pull request by pauldreik**
This fixes #562 

A commit with git blame ignore should be added, once we know the hash of the big reformat commit (currently c44bbc9813c99c427bb246ef099c8997d615930c).

I pushed the docker image to my personal package repository, let's discuss if I or @lemire can push it directly to simdutf instead so the project is more self contained.

### CODE REVIEW

File: src/simdutf/icelake/begin.h [hunk 1]
```diff
@@ -6,6 +6,7 @@
 SIMDUTF_TARGET_ICELAKE
 #endif
 
-#if SIMDUTF_GCC11ORMORE // workaround for https://gcc.gnu.org/bugzilla/show_bug.cgi?id=105593
-SIMDUTF_DISABLE_GCC_WARNING(-Wmaybe-uninitialized)
-#endif // end of workaround
\ No newline at end of file
+#if SIMDUTF_GCC11ORMORE // workaround for
+                        // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=105593
+SIMDUTF_DISABLE_GCC_WARNING(-Wmaybe - uninitialized)
```
Comment (lemire) [inline src/simdutf/icelake/begin.h:11]:
```suggestion
SIMDUTF_DISABLE_GCC_WARNING(-Wmaybe-uninitialized)
```

**Comment by lemire** [comment]
I would be happy to merge this.

@FranciscoThiesen : can you have a look? I know that you might have been working on something and I would rather we do not create conflicts from h*ll.


It is important that anyone working on the code has a look at this PR as it changes many files and may cause difficult-to-resolve conflicts.

**Comment by lemire** [comment]
@anonrig  @FranciscoThiesen 

This PR would reformat everything. We are waiting for a strategy timing to fire it up.

**Review by anonrig — APPROVED** [review]

**Comment by lemire** [comment]
Ok. I am merging. @FranciscoThiesen and @anonrig ... please be careful with conflicts.

**Comment by lemire** [comment]
@pauldreik Done. :-)

**Pull request merged**